Packages/com.unity.inputsystem/InputSystem/Plugins/InputForUI/InputSystemProvider.cs

```diff
@@ -0,0 +1,679 @@
+
+namespace UnityEngine.InputSystem.Plugins.InputForUI
+{
+    using Event = UnityEngine.InputForUI.Event;
+    using EventModifiers = UnityEngine.InputForUI.EventModifiers;
+    using EventProvider = UnityEngine.InputForUI.EventProvider;
+
+    internal class InputSystemProvider : IEventProviderImpl
+    {
+        InputEventPartialProvider m_InputEventPartialProvider;
+
+        Configuration m_Cfg;
+
+        InputActionAsset m_InputActionAsset;
+        InputActionReference m_PointAction;
+        InputActionReference m_MoveAction;
+        InputActionReference m_SubmitAction;
+        InputActionReference m_CancelAction;
+        InputActionReference m_LeftClickAction;
+        InputActionReference m_MiddleClickAction;
+        InputActionReference m_RightClickAction;
+        InputActionReference m_ScrollWheelAction;
+
+        InputAction m_NextPreviousAction;
+
+        List<Event> m_Events = new List<Event>();
+
+        PointerState m_MouseState;
+
+        PointerState m_PenState;
+        bool m_SeenPenEvents;
+
+        PointerState m_TouchState;
+        bool m_SeenTouchEvents;
+
+        const float k_SmallestReportedMovementSqrDist = 0.01f;
+
+        NavigationEventRepeatHelper m_RepeatHelper = new();
+        bool m_ResetSeenEventsOnUpdate;
+
+        static InputSystemProvider()
+        {
+            // Only if InputSystem is enabled in the PlayerSettings do we set it as the provider.
+            // This includes situations where both InputManager and InputSystem are enabled.
+#if ENABLE_INPUT_SYSTEM
+            EventProvider.SetInputSystemProvider(new InputSystemProvider());
+#endif
+        }
+
+        [RuntimeInitializeOnLoadMethod(loadType: RuntimeInitializeLoadType.SubsystemRegistration)]
+        static void Bootstrap() {} // Empty function. Exists only to invoke the static class constructor in Runtime Players
+
+        EventModifiers m_EventModifiers => m_InputEventPartialProvider._eventModifiers;
+
+        DiscreteTime m_CurrentTime => (DiscreteTime)Time.timeAsRational;
+
+        const uint k_DefaultPlayerId = 0u;
+
+        public void Initialize()
+        {
+            m_InputEventPartialProvider ??= new InputEventPartialProvider();
+            m_InputEventPartialProvider.Initialize();
+
+            m_Events.Clear();
+
+            m_MouseState.Reset();
+
+            m_PenState.Reset();
+            m_SeenPenEvents = false;
+
+            m_TouchState.Reset();
+            m_SeenTouchEvents = false;
+
+            // TODO should UITK somehow override this?
+            m_Cfg = Configuration.GetDefaultConfiguration();
+            RegisterActions(m_Cfg);
+        }
+
+        public void Shutdown()
+        {
+            UnregisterActions(m_Cfg);
+
+            m_InputEventPartialProvider.Shutdown();
+            m_InputEventPartialProvider = null;
+        }
+
+        public void Update()
+        {
+            m_InputEventPartialProvider.Update();
+
+            m_Events.Sort(SortEvents);
```

Comment:
Would be handy to have a comment to clarify what we are sorting on - seems to be type so so the loop below is ordered correctly.